src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs

```diff
@@ -206,16 +207,44 @@ internal static IEnumerable<DownloadDependency> GetPackageDownloads(IMSBuildProj
             }
         }
 
+        /// <summary>
+        /// Gets the centrally defined package version information.
+        /// </summary>
+        /// <param name="project">The <see cref="ProjectInstance" /> to get PackageVersion for.</param>
+        /// <param name="skippedPackageIds">An <see cref="IEnumerable{string}" /> of package ids that were already added as package references. These will be excluded from the resulted list.</param>
+        /// <returns>An <see cref="IEnumerable{CentralVersionDependency}" /> containing the package versions for the specified project.</returns>
+        internal static IEnumerable<CentralVersionDependency> GetCentralPackageVersions(IMSBuildProject project, IEnumerable<string> skippedPackageIds)
+        {
+            var skippedPackageIdsHashSet = skippedPackageIds.Select(id=>id.ToLowerInvariant()).ToHashSet();
+            var packageVersionItems = GetDistinctItemsOrEmpty(project, "PackageVersion")
+                .Where(item=>!skippedPackageIdsHashSet.Contains(item.Identity.ToLowerInvariant()))
+                .ToList();
+
+            foreach (var projectItemInstance in packageVersionItems)
+            {
+                string id = projectItemInstance.Identity;
+                string version = projectItemInstance.GetProperty("Version");
+                VersionRange versionRange = !string.IsNullOrWhiteSpace(version) ? VersionRange.Parse(version) : VersionRange.All;
+
+                yield return new CentralVersionDependency(id, versionRange);
+            }
+        }
+
         /// <summary>
         /// Gets the package references for the specified project.
         /// </summary>
         /// <param name="project">The <see cref="ProjectInstance" /> to get package references for.</param>
+        /// <param name="cpvmEnabled">A flag for cpvmEnabled status.</param>
         /// <returns>A <see cref="List{LibraryDependency}" /> containing the package references for the specified project.</returns>
-        internal static List<LibraryDependency> GetPackageReferences(IMSBuildProject project)
+        internal static List<LibraryDependency> GetPackageReferences(IMSBuildProject project, bool cpvmEnabled)
         {
             // Get the distinct PackageReference items, ignoring duplicates
-            var packageReferenceItems = project.GetItems("PackageReference").Distinct(ProjectItemInstanceEvaluatedIncludeComparer.Instance).ToList();
-
+            var packageReferenceItems = GetDistinctItemsOrEmpty(project, "PackageReference").ToList();
+            if(cpvmEnabled)
+            {
+                MergeCentralPackageVersionsInPackageReferences(packageReferenceItems, GetDistinctItemsOrEmpty(project, "PackageVersion").ToList());
```

Comment:
You're getting the list of PackageVersion items twice.  I would expect the merging of PackageReference and PackageVersion to happen in the API after you've populated the DG spec.  But instead you're doing the merging here which seems odd.  Is it not possible to make the DG spec not know about these things and instead have the API that does the restore figure out what versions should be applied?

Reply:
Moved the merge operation.